Corona-Warn-App/src/main/java/de/rki/coronawarnapp/playbook/Playbook.kt

```diff
@@ -19,23 +32,210 @@ import de.rki.coronawarnapp.server.protocols.internal.pt.CheckInOuterClass
+    private val uid = UUID.randomUUID().toString()
+    private val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
 
-    suspend fun submit(data: SubmissionData)
+    suspend fun initialRegistration(
+        tokenRequest: RegistrationRequest
+    ): RegistrationData {
+        Timber.i("[$uid] New Initial Registration Playbook")
 
-    suspend fun dummy()
+        // real registration
+        val (registrationToken, registrationException) = executeCapturingExceptions {
+            verificationServer.retrieveRegistrationToken(tokenRequest)
+        }
+
+        // if the registration succeeded continue with the real test result retrieval
+        // if it failed, execute a dummy request to satisfy the required playbook pattern
+        val (testResultResponse, testResultException) = if (registrationToken != null) {
+            executeCapturingExceptions { verificationServer.pollTestResult(registrationToken) }
+        } else {
+            ignoreExceptions { verificationServer.retrieveTanFake() }
+            null to null
+        }
+
+        // fake submission
+        ignoreExceptions { submissionServer.submitFakePayload() }
+
+        coroutineScope.launch { followUpPlaybooks() }
+
+        // if registration and test result retrieval succeeded, return the result
+        if (registrationToken != null && testResultResponse != null) {
+
+            return RegistrationData(
+                registrationToken = registrationToken,
+                testResultResponse = testResultResponse,
+            )
+        }
+
+        // else propagate the exception of either the first or the second step
+        propagateException(registrationException, testResultException)
+    }
+
+    suspend fun testResult(registrationToken: String): CoronaTestResultResponse {
+        Timber.i("[$uid] New Test Result Playbook")
+
+        // real test result
+        val (testResult, exception) =
+            executeCapturingExceptions { verificationServer.pollTestResult(registrationToken) }
+
+        // fake verification
+        ignoreExceptions { verificationServer.retrieveTanFake() }
+
+        // fake submission
+        ignoreExceptions { submissionServer.submitFakePayload() }
+
+        coroutineScope.launch { followUpPlaybooks() }
+
+        return testResult ?: propagateException(exception)
+    }
+
+    suspend fun retrieveTan(registrationToken: String, authCode: String?): String {
+        return if (authCode == null) {
+            // real auth code
+            val (newAuthCode, exception) = executeCapturingExceptions {
+                verificationServer.retrieveTan(registrationToken)
+            }
+            if (newAuthCode == null) {
+                propagateException(wrapException(exception))
+            }
+            newAuthCode
+        } else {
+            // fake request
+            verificationServer.retrieveTanFake()
+            authCode
+        }
+    }
+
+    suspend fun submit(
+        data: SubmissionData?
```

Comment:
I would split this into two methods one that submits a non nullable data and one for fake submission